Revise the manual (#3231)
GaloisInc/saw-script: doc/saw-user-manual/symbolic-execution.md

# Introduction to Symbolic Execution

> Analysis of Java and LLVM in SAW relies heavily on *symbolic

Reviewer: > Analysis of Java and LLVM

and Rust!

Author: Fixed

> execution*, so some background on how this process works can help with

Reviewer: I think we should also state here that _symbolic simulation_ (or just _simulation_) is a common synonym for symbolic execution.

Author: Fixed

understanding the system and its behavior.
If you are already reasonably familiar with symbolic execution and its
concepts and limitations, you can safely skip this chapter.

At the most abstract level, symbolic execution works like normal
program execution except that program values can be *symbolic*:
instead of a single concrete value like `3`, a value can be a symbol
that stands for any value of the proper type, or a subset of those
values, or an expression in terms of such symbols.
Typically we create fresh symbols to serve as the program input, and

> then as symbolic execution executes, new values are computed from old

Reviewer: "execution executes" is a somewhat awkward way to phrase this. Perhaps rephrase this as "as symbolic execution proceeds"?

Author: Indeed. fixed

}
:::

If you call this function with two concrete inputs and assign the
result, like this:

:::{code-block} c
int r = max(5, 4);
:::

then it will assign the value `5` to `r`.
Now, however, we can create two fresh symbols `a` and `b` and call the
function with them as the arguments:

:::{code-block} c
int r = max(a, b);
:::

The symbolic value that we store in `r` is some form of the expression
`if b > a then b else a`, possibly specifically the C expression

> `b > a ? b : a` and possibly something slightly different.

Reviewer: Minor word-smithing to avoid two adverbs in a row:

```suggestion
`if b > a then b else a`, possibly the specific C expression
`b > a ? b : a` and possibly something slightly different.
```

Author: Fixed (slightly differently)

(Aside: whether symbolic expressions inside a symbolic execution
engine are represented in the input language or some other language is
a design decision.
The tradeoffs involved are well beyond the scope of this introduction.)

To generate this expression, however, you will note that we have to
execute both branches of the `if` statement in the original function.
Any concrete execution can evaluate the `if` condition to either 0 or
1 and then execute one branch of the if and not the other.
However, during symbolic execution the if condition is the symbolic
expression `b > a`, which could be either true or false.
Therefore we need to proceed down both branches of the if.
This in turn produces (in general) two different program states
afterward, which then need to be merged back together.
This merge applies the if condition to the results of each side, and

> introduces the if that appears in our symbolic result.

Reviewer: For consistency with the first part of this paragraph, which renders `if` in monospace font:

```suggestion
However, during symbolic execution the `if` condition is the symbolic
expression `b > a`, which could be either true or false.
Therefore we need to proceed down both branches of the `if`.
This in turn produces (in general) two different program states
afterward, which then need to be merged back together.
This merge applies the `if` condition to the results of each side, and
introduces the `if` that appears in our symbolic result.
```

Author: Fixed (as well as the one on line 61)

You will have noticed that the symbolic expression we get from calling
`max` with arbitrary inputs is scarcely different from the original
code; this is because `max` is a very simple function and we have used
the most simple possible input context.

## Path Satisfiability

Suppose, however, that we already knew something about the inputs before
calling the function.
Consider this calling context instead:

:::{code-block} c
int minmax(unsigned int x, unsigned int y) {
    int r;

    if (x > y) {
        r = max(x, y);
    } else {
        r = min(x, y);
    }

>     return r;

Reviewer 2: ```suggestion
    return r;
}
```

:::

(This is perhaps a silly function to write, but it will serve as an
example.)

If in turn we call this with symbolic inputs `a` and `b`, when we
get to `max` we know that `a > b`, because we're inside the first
branch of the `if` in `minmax`.
This expression (that tells us what must be true to get where we
are) is called the *path condition* because it's associated with
the execution path that we took to get to the current program point.

Now when we reach the `if` in `max`, it checks `b > a`.
We can conclude that this is false: there is no satisfying assignment
that makes `a > b && b > a` true.
We can therefore skip the true branch of the `if` entirely; the
symbolic return value we get from `max` is just `a`.

> What we have just done is called *path satisifiability checking*:

Reviewer 2: ```suggestion
What we have just done is called *path satisfiability checking*:
```

When we reach the condition of a loop, unless we can determine that
the condition is definitely false, we generate two cases: one where
the loop ends and one where we go around the loop again.
This means that unless the symbolic program state evolves to a
configuration where the condition is definitely false, we will keep
generating more cases where we go around the loop again, and do so
forever.

Loops that reach a definitely false state are said to *symbolically
terminate*.
For example, loops with concrete, constant bounds symbolically
terminate: if the count starts at 0, and goes up by one on each
loop iteration, and the condition is, say, `i < 15`, after unrolling
the loop fifteen times `i` will be `15` and we stop.

Other loops may also symbolically terminate.
The following contrived example symbolically terminates, because as
it goes it accumulates (in its path condition) assertions about
successive values of `i` not being zero mod 8.
Because 5 and 8 are relatively prime, it will eventually accumulate
all 8 possible values, of which one must be zero; at that point there
is no satisfying assignment for `i` that continues the loop (this is

> within the ability of a solver to discover) and we stop.

Reviewer: It might be worth mentioning that it is within the ability of a solver to discover if path satisfiability checking is enabled.

Author: I took that out because I didn't want to get into path satisfiability checking here; turning it off is an advanced topic and people shouldn't be thinking about it at this stage. I put a reference back in, except that there's no advanced topics chapter to refer to yet...

(In program analysis terms, symbolic execution is *precise* but not
*complete*.)

There are several approaches commonly taken in practice to approximate
the behavior of unbounded loops and still get useful proof results.
SAW's support for these is discussed in later chapters.

<!--
   XXX: this section used to mention the effects of turning path
   satisfiability checking on:
   
   : Normally, most of the Java and LLVM analysis commands simply compare
   : branch conditions to the constant `True` or `False` to determine whether
   : a branch may be feasible. However, each form of analysis allows branch
   : satisfiability checking to be turned on if needed, in which case
   : functions like `f` above will terminate.

   This needs to get stuffed in somewhere later. It also needs to be
   recast as turning path satisfiability _off_, because "normally" it
   should be on. Turning it off is an advanced feature and you need to
   know what you're doing.

   (Though now we've at least introduced what path satisfiability
   checking _is_, which was missing before.)
-->

## Setting up Symbolic Execution

A symbolic execution system is very similar to an interpreter, just

> one has a different notion of what constitutes a value and executes

Reviewer 2: ```suggestion
A symbolic execution system is very similar to an interpreter, but with
a different notion of what constitutes a value and it executes
```
